### PR TITLE
refactor(storage): refactor read_filter_for_batch to avoid continuous locking

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by wenym1 — APPROVED** [review]
LGTM

File: src/storage/src/hummock/state_store.rs [hunk 1]
```diff
@@ -136,7 +136,26 @@ impl HummockStorage {
                 if read_version_vec.is_empty() {
                     (Vec::default(), Vec::default(), (**pinned_version).clone())
                 } else {
-                    read_filter_for_batch(epoch, table_id, key_range, read_version_vec)?
+                    let mut staging_vec = Vec::with_capacity(read_version_vec.len());
+                    let mut max_mce = 0;
+                    for read_version in &read_version_vec {
+                        let read_version_guard = read_version.read();
+                        staging_vec.push(read_version_guard.staging().clone());
+                        max_mce = std::cmp::max(
```
Comment (zwang28) [inline src/storage/src/hummock/state_store.rs:144]: Will `staging().clone()` be expensive?

**Review by StrikeW — APPROVED** [review]
LGTM

File: src/storage/src/hummock/state_store.rs (hunk 1 above)
Comment (Li0k) [reply src/storage/src/hummock/state_store.rs]: refactor it with prune, PTAL

File: src/storage/src/hummock/store/version.rs [hunk 2]
```diff
@@ -178,12 +176,8 @@ impl StagingVersion {
             .sst
             .iter()
             .filter(move |staging_sst| {
-                let sst_min_epoch = *staging_sst.epochs.first().expect("epochs not empty");
                 let sst_max_epoch = *staging_sst.epochs.last().expect("epochs not empty");
-                assert!(
-                    sst_max_epoch <= min_epoch_exclusive || sst_min_epoch > min_epoch_exclusive
```
Comment (hzxa21) [inline src/storage/src/hummock/store/version.rs:184]: Can we add back this assert in `read_filter_for_batch`?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM

**Review by zwang28 — APPROVED** [review]

**Pull request merged**